### PR TITLE
Full fledged ODS-CORE Tests

### PR DESCRIPTION
**Pull request by hugowschneider**
This is a work in progress based on PR #342. It is the implementation of #343 which deploys a new project in OpenShift using `oc cluster up` and `opendevstack/mockbucket`.

If this works, this will be the base infrastructure for testing `opendevstack/ods-quickstarters`.

### CODE REVIEW

File: tests/create-projects/jenkinsfile_test.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+				stdout,
+				stderr)
+		} else {
+			t.Fatalf(
+				"Error during build: \nStdOut: %s\nStdErr: %s",
+				stdout,
+				stderr)
+		}
+
+	}
```
Comment (michaelsauter) [inline tests/create-projects/jenkinsfile_test.go]: Could we extract these calls/checks into a method? Would make it more readable

File: tests/create-projects/jenkinsfile_test.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+	if err = utils.FindProject(projects, utils.PROJECT_NAME_CD); err != nil {
+		t.Fatal(err)
+	}
+	if err = utils.FindProject(projects, utils.PROJECT_NAME_TEST); err != nil {
+		t.Fatal(err)
+	}
+	if err = utils.FindProject(projects, utils.PROJECT_NAME_DEV); err != nil {
+		t.Fatal(err)
+	}
```
Comment (michaelsauter) [inline tests/create-projects/jenkinsfile_test.go]: Put into a loop

File: tests/create-projects/jenkinsfile_test.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+	roleBindings, _ = rbacV1Client.RoleBindings(utils.PROJECT_NAME_TEST).List(metav1.ListOptions{})
+	if err = utils.FindRoleBinding(roleBindings, "default", "ServiceAccount", utils.PROJECT_NAME_TEST, "system:image-builder"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = utils.FindRoleBinding(roleBindings, "jenkins", "ServiceAccount", utils.PROJECT_NAME_CD, "admin"); err != nil {
+		t.Fatal(err)
+	}
```
Comment (michaelsauter) [inline tests/create-projects/jenkinsfile_test.go]: Table test for these please

File: tests/ods-setup/setup-ods-project_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		t.Fatal(err)
+	}
+
+	if err = utils.FindImageTag(images, "jenkins-slave-base", "latest"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = utils.FindImageTag(images, "jenkins-webhook-proxy", "latest"); err != nil {
+		t.Fatal(err)
+	}
```
Comment (michaelsauter) [inline tests/ods-setup/setup-ods-project_test.go]: Table test please

File: tests/scripts/deploy-mocks.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+   -v|--verbose) set -x;;
+
+   -h|--help) usage; exit 0;;
+
+   *) echo "Unknown parameter passed: $1"; usage; exit 1;;
+ esac; shift; done
+
+URL=$(oc config view --minify -o jsonpath='{.clusters[*].cluster.server}')
+if [ ${URL} != "https://127.0.0.1:8443" ]; then
+    echo "You are not in a local cluster. Stopping now!!!"
```
Comment (michaelsauter) [inline tests/scripts/deploy-mocks.sh]: I like :D

File: tests/scripts/deploy-mocks.sh [hunk 6]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
```
Comment (michaelsauter) [inline tests/scripts/deploy-mocks.sh]: Maybe some comments what this does?

File: tests/scripts/deploy-mocks.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+fi
+
+
+source ${BASH_SOURCE%/*}/../../ods-config/ods-core.env
+docker run -d -p "8080:8080" \
+           --env="BASIC_USERNAME=${CD_USER_ID}" \
+           --env="BASIC_PASSWORD=${CD_USER_PWD}" \
+           --env="REPOS=opendevstack/ods-core.git;opendevstack/ods-configuration.git" \
+           --name mockbucket \
+           hugowschneider/mockbucket:latest
```
Comment (michaelsauter) [inline tests/scripts/deploy-mocks.sh]: newline please

File: tests/scripts/deploy-mocks.sh (hunk 6 above)
Comment (michaelsauter) [inline tests/scripts/deploy-mocks.sh]: `set -e` would be good

File: tests/scripts/recreate-test-infrastructure.sh [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+
+URL=$(oc config view --minify -o jsonpath='{.clusters[*].cluster.server}')
+
+if [ ${URL} != "https://127.0.0.1:8443" ]; then
+    echo "You are not in a local cluster. Stopping now!!!"
+fi
+
+if ! oc get clusterroles | grep request_role; then
+echo "{ \"apiVersion\": \"v1\", \"kind\": \"ClusterRole\", \"metadata\": { \"name\": \"request_role\" }, \"rules\": [ { \"verbs\": [\"create\"], \"resources\": [\"projectrequests\"] }, { \"verbs\": [\"get\",\"list\",\"watch\"], \"resources\": [\"*\"] } ] }" \
```
Comment (michaelsauter) [inline tests/scripts/recreate-test-infrastructure.sh]: could be a json file in the same folder

File: tests/scripts/recreate-test-infrastructure.sh [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+fi
+
+if oc project ${NAMSPACE}; then
+    oc delete project ${NAMSPACE}
+fi
+
+${BASH_SOURCE%/*}/../../ods-setup/setup-ods-project.sh --verbose --force --namespace ${NAMSPACE}
+
+${BASH_SOURCE%/*}/deploy-mocks.sh  --verbose
+sleep 10 # Waiting for service to boot up
```
Comment (michaelsauter) [inline tests/scripts/recreate-test-infrastructure.sh]: Hmm. Should this rather be in `deploy-mocks`? Maybe as a param `--wait`?

File: tests/scripts/recreate-test-infrastructure.sh [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    ${BASH_SOURCE%/*}/../../create-projects/create-projects.sh  --verbose
+
+PROJECT_ID=${PROJECT_ID} \
+CD_USER_TYPE=general \
+CD_USER_ID_B64=${CD_USER_ID_B64} \
+PIPELINE_TRIGGER_SECRET=${PIPELINE_TRIGGER_SECRET_B64} \
+    ${BASH_SOURCE%/*}/../../create-projects/create-cd-jenkins.sh --ods-namespace ${NAMSPACE} --force --verbose
+
+oc adm policy add-cluster-role-to-user self-provisioner system:serviceaccount:prov-cd:jenkins
+
```
Comment (michaelsauter) [inline tests/scripts/recreate-test-infrastructure.sh]: newline

File: tests/scripts/setup-mocked-ods-repo.sh [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+git -c http.sslVerify=false push mockbucket --set-upstream "${HEAD}:${REF}"
+git remote remove mockbucket
+
+mkdir -p "${HOME}/ods-configuration"
+git -C "${HOME}/ods-configuration" init
+cp ${BASH_SOURCE%/*}/../../ods-config/ods-core.env ${HOME}/ods-configuration
+git -C "${HOME}/ods-configuration" add ods-core.env
+git -C "${HOME}/ods-configuration" commit -m "Initial Commit"
+git -C "${HOME}/ods-configuration" remote add mockbucket http://$(urlencode ${CD_USER_ID}):$(urlencode ${CD_USER_PWD})@${BITBUCKET_HOST}/scm/opendevstack/ods-configuration.git
+git -C "${HOME}/ods-configuration" -c http.sslVerify=false push mockbucket --set-upstream "$(git -C "${HOME}/ods-configuration" rev-parse --abbrev-ref HEAD):${REF}"
```
Comment (michaelsauter) [inline tests/scripts/setup-mocked-ods-repo.sh]: Doing a `cd ${HOME}/ods-configuration` beforehand would make the lines easier to read ... no?

**Comment by hugowschneider** [comment]
I am blocked by Travis CI memory limitations. 7.5GB is not enough :(. Need some light here...

**Comment by michaelsauter** [comment]
@hugowschneider If I understand correctly, your automated tests worked without Jenkins. I would suggest we commit these then, and leave the Jenkins stuff in the code, but without running it in Travis. That way, we have at least some automated tests, and the possibility to run the other automated tests at regular intervals against a custom machine. We then need to think about how to get more memory, maybe we find a way.

**Comment by hugowschneider** [comment]
PR #342 runs only scripts and this PR does both, Jenkinsfile and scripts. I like the idea, not ideal but I don't see a way around resources in TravisCI or Github Actions

**Comment by clemensutschig** [comment]
@hugowschneider .. where is the Limit poppig up?

**Comment by hugowschneider** [comment]
@clemensutschig, I am not sure. OpenShift API just stops responding during the tests. I had the same behavior in a VM and it stopped when I increased memory. 

I will investigate it further... to see also if there is a way around

**Comment by clemensutschig** [comment]
Hmmm what about using Another Cloud für the oc cluster up.. to save Resources.. kg rephrase.. how is ocp testing in github done.. ? Just a bunch questions

**Comment by hugowschneider** [comment]
@clemensutschig, I do not know where the problem occurs. The tests run a while and suddenly OC API stops replying. We can see it in the build https://travis-ci.com/opendevstack/ods-core/jobs/273003156 

I had the same issue when using a VM and it stopped when I increased memory. I am now using a 64Gb machine and looking into the OpenShift processes, there is one OpenShift process that alone uses 12Gb. There are also docker processes that eat up memory.

We could use another cloud but the problem is storing the credentials and licenses. Travis CI can store it securely but it does not run any build/test on PR from forks, only in PRs from the same repo (https://docs.travis-ci.com/user/pull-requests#pull-requests-and-security-restrictions).

For how OCP test its code. My guess, by looking into the PR's, is that they use some actions using an OpenShift CI Robot (https://github.com/openshift-ci-robot).

Following some links, I could also find https://github.com/openshift/ci-tools, which I guess has the implementation of this robot.

Lots of guesswork here :).

**Comment by clemensutschig** [comment]
Hmmm can you Check the cluster .. how many Master and Worker Nodes? We just Need One..  and Maybe there is a way to control that.. or make them rather small..

**Comment by hugowschneider** [comment]
It is for sure a single node cluster.

**Comment by hugowschneider** [comment]
I tried to disable the web-console to save some memory but the problem is still there

**Comment by hugowschneider** [comment]
The good news is: I could run everything in a 4Gb machine in AWS. 

Now I need to figure out what is the problem in TravisCI

**Comment by clemensutschig** [comment]
Same error? Is the Process blocking.. or what do you get?

hugowschneider <notifications@github.com> schrieb am Mi., 8. Jän. 2020,
11:45:

File: jenkins/ocp-config/bc.yml [hunk 12]
```diff
@@ -14,9 +14,7 @@ objects:
     postCommit: {}
     resources: {}
     runPolicy: Serial
-    triggers:
-      - imageChange: {}
-        type: ImageChange
+    triggers: []
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply jenkins/ocp-config/bc.yml]: I created an issue for this one... I still need to create the PR

File: tests/scripts/setup-mocked-ods-repo.sh (hunk 11 above)
Comment (hugowschneider) [reply tests/scripts/setup-mocked-ods-repo.sh]: Done!

File: tests/scripts/recreate-test-infrastructure.sh (hunk 10 above)
Comment (hugowschneider) [reply tests/scripts/recreate-test-infrastructure.sh]: Done!

File: tests/scripts/recreate-test-infrastructure.sh (hunk 9 above)
Comment (hugowschneider) [reply tests/scripts/recreate-test-infrastructure.sh]: Added --wait

File: tests/scripts/recreate-test-infrastructure.sh (hunk 8 above)
Comment (hugowschneider) [reply tests/scripts/recreate-test-infrastructure.sh]: Done

File: tests/scripts/deploy-mocks.sh (hunk 6 above)
Comment (hugowschneider) [reply tests/scripts/deploy-mocks.sh]: added set -ue

File: ods-setup/setup-jenkins-images.sh [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+   *) echo "Unknown parameter passed: $1"; usage; exit 1;;
+ esac; shift; done
+
+if ! oc whoami; then
+  echo "You should be logged to run the script"
+  exit 1
+fi
+
+if ! oc project ${NAMESPACE}; then
+  echo "The project '${NAMESPACE}' does not exists. Please setup the project using 'setup-ods-project.sh'"
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-jenkins-images.sh]: Done

File: ods-setup/setup-jenkins-images.sh [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+   -r|--ods-base-repository) REPOSITORY="$2"; shift;;
+
+   -b=*|--ods-ref=*) REF="${1#*=}";;
+   -b|--ods-ref) REF="$2"; shift;;
+
+   *) echo "Unknown parameter passed: $1"; usage; exit 1;;
+ esac; shift; done
+
+if ! oc whoami; then
+  echo "You should be logged to run the script"
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-jenkins-images.sh]: Done

File: ods-setup/setup-jenkins-images.sh [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+#!/usr/bin/env bash
+set -xe
+
+function usage {
+   printf "usage: %s [options]\n", $0
+   printf "\t--force\tIgnores warnings and error with tailor --force\n"
+   printf "\t-h|--help\tPrints the usage\n"
+   printf "\t-v|--verbose\tVerbose output\n"
+   printf "\t-t|--tailor\tChanges the executable of tailor. Default: tailor\n"
+   printf "\t-n|--namespace\tChanges the default namespace. Default: cd\n"
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-jenkins-images.sh]: added better documentation in usage

File: jenkins/ocp-config/bc.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (hugowschneider) [reply jenkins/ocp-config/bc.yml]: PR is there

File: jenkins/ocp-config/Tailorfile [hunk 16]
```diff
@@ -1,6 +1,5 @@
-namespace cd
 selector app=jenkins
-param-file ../../../ods-configuration/ods-core.env
+param-file ../../ods-config/ods-core.env
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply jenkins/ocp-config/Tailorfile]: I need to double check what does mean for the test... still on going

File: ods-setup/setup-ods-project.sh [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+   printf "\t-t|--tailor\tChanges the executable of tailor. Default: tailor\n"
+   printf "\t-n|--namespace\tChanges the default namespace. Default: cd\n"
+
+}
+TAILOR="tailor"
+NAMESPACE="cd"
+
+while [[ "$#" -gt 0 ]]; do case $1 in
+
+   -v|--verbose) set -x;;
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-ods-project.sh]: This is fixed...

File: ods-setup/setup-jenkins-images.sh [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+   printf "\t-b|--ods-ref\tODS reference in repository. Overrides default in settings file\n"
+
+}
+TAILOR="tailor"
+NAMESPACE="cd"
+REPOSITORY=""
+REF=""
+while [[ "$#" -gt 0 ]]; do case $1 in
+
+   -v|--verbose) set -x;;
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-jenkins-images.sh]: Fixed

File: tests/create-projects/create-projects_test.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    roleBindings, _ = rbacV1Client.RoleBindings(utils.PROJECT_NAME_TEST).List(metav1.ListOptions{})
+    if err = utils.FindRoleBinding(roleBindings, "default", "ServiceAccount", utils.PROJECT_NAME_TEST, "system:image-builder"); err != nil {
+        t.Fatal(err)
+    }
+
+    if err = utils.FindRoleBinding(roleBindings, "jenkins", "ServiceAccount", utils.PROJECT_NAME_CD, "admin"); err != nil {
+        t.Fatal(err)
+    }
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-projects_test.go]: Done

File: tests/create-projects/jenkinsfile_test.go (hunk 2 above)
Comment (hugowschneider) [reply tests/create-projects/jenkinsfile_test.go]: Done

File: tests/ods-setup/setup-ods-project_test.go (hunk 4 above)
Comment (hugowschneider) [reply tests/ods-setup/setup-ods-project_test.go]: Done

File: tests/scripts/deploy-mocks.sh (hunk 5 above)
Comment (hugowschneider) [reply tests/scripts/deploy-mocks.sh]: :)

File: tests/scripts/deploy-mocks.sh (hunk 7 above)
Comment (hugowschneider) [reply tests/scripts/deploy-mocks.sh]: I guess all newlines are there...

File: tests/scripts/deploy-mocks.sh (hunk 6 above)
Comment (hugowschneider) [reply tests/scripts/deploy-mocks.sh]: Done

File: tests/create-projects/jenkinsfile_test.go (hunk 3 above)
Comment (hugowschneider) [reply tests/create-projects/jenkinsfile_test.go]: Done

File: ods-setup/setup-jenkins-images.sh [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+${TAILOR} update ${FORCE} --context-dir=${BASH_SOURCE%/*}/../jenkins/ocp-config --non-interactive -n ${NAMESPACE} ${REF} ${REPOSITORY}
+
+echo "Start Jenkins Builds"
+oc start-build -n ${NAMESPACE} jenkins-master --follow
+oc start-build -n ${NAMESPACE} jenkins-slave-base --follow
+oc start-build -n ${NAMESPACE} jenkins-webhook-proxy --follow
+
+oc tag ${NAMESPACE}/jenkins-master:test ${NAMESPACE}/jenkins-master:latest
+oc tag ${NAMESPACE}/jenkins-slave-base:test ${NAMESPACE}/jenkins-slave-base:latest
+oc tag ${NAMESPACE}/jenkins-webhook-proxy:test ${NAMESPACE}/jenkins-webhook-proxy:latest
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-jenkins-images.sh]: Like I said... I guess all new lines are there

File: ods-setup/setup-ods-project.sh [hunk 21]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
+set -xe
+
+function usage {
+   printf "usage: %s [options]\n", $0
+   printf "\t--force\tIgnores warnings and error with tailor --force\n"
+   printf "\t-h|--help\tPrints the usage\n"
+   printf "\t-v|--verbose\tVerbose output\n"
+   printf "\t-t|--tailor\tChanges the executable of tailor. Default: tailor\n"
+   printf "\t-n|--namespace\tChanges the default namespace. Default: cd\n"
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply ods-setup/setup-ods-project.sh]: Added documentation why 2 different names

File: tests/create-projects/create-cd-jenkins_test.go [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+	user := base64.StdEncoding.EncodeToString([]byte("myuser"))
+	//secret := base64.StdEncoding.EncodeToString([]byte("mysecret"))
+	stdout, stderr, err := utils.RunScriptFromBaseDir("create-projects/create-cd-jenkins.sh", []string{"--force", "--verbose"},
+		utils.PROJECT_ENV_VAR,
+		"CD_USER_TYPE=general",
+		fmt.Sprintf("CD_USER_ID_B64=%s", user))
+	// fmt.Sprintf("PIPELINE_TRIGGER_SECRET=%s",secret)	)
+	if err == nil {
+		t.Fatalf(
+			"Execution of `create-cd-jenkins.sh` must fail if no PROJECT_ID is set: \nStdOut: %s\nStdErr: %s",
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-cd-jenkins_test.go]: Fixed

File: tests/create-projects/create-cd-jenkins_test.go [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+	user := base64.StdEncoding.EncodeToString([]byte("myuser"))
+	secret := base64.StdEncoding.EncodeToString([]byte("mysecret"))
+	stdout, stderr, err := utils.RunScriptFromBaseDir("create-projects/create-cd-jenkins.sh", []string{"--force", "--verbose"},
+		utils.PROJECT_ENV_VAR,
+		// "CD_USER_TYPE=general",
+		fmt.Sprintf("CD_USER_ID_B64=%s", user),
+		fmt.Sprintf("PIPELINE_TRIGGER_SECRET=%s", secret))
+	if err == nil {
+		t.Fatalf(
+			"Execution of `create-cd-jenkins.sh` must fail if no PROJECT_ID is set: \nStdOut: %s\nStdErr: %s",
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-cd-jenkins_test.go]: Fixed

File: tests/create-projects/create-cd-jenkins_test.go [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+		"CD_USER_TYPE=general",
+		fmt.Sprintf("CD_USER_ID_B64=%s", user))
+	// fmt.Sprintf("PIPELINE_TRIGGER_SECRET=%s",secret)	)
+	if err == nil {
+		t.Fatalf(
+			"Execution of `create-cd-jenkins.sh` must fail if no PROJECT_ID is set: \nStdOut: %s\nStdErr: %s",
+			stdout,
+			stderr)
+	}
+}
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-cd-jenkins_test.go]: Done

File: tests/create-projects/create-cd-jenkins_test.go [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestCreateJenkins(t *testing.T) {
+	_ = utils.RemoveAllTestOCProjects()
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-cd-jenkins_test.go]: Done

File: tests/create-projects/create-cd-jenkins_test.go [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+	odsNamespace := "ods"
+	stdout, stderr, err := utils.RunScriptFromBaseDir("create-projects/create-projects.sh", []string{}, utils.PROJECT_ENV_VAR)
+	if err != nil {
+		t.Fatalf(
+			"Execution of `create-project.sh` failed: \nStdOut: %s\nStdErr: %s",
+			stdout,
+			stderr)
+	}
+
+	values, err := utils.ReadValues()
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-cd-jenkins_test.go]: Done

File: tests/create-projects/jenkinsfile_test.go (hunk 1 above)
Comment (hugowschneider) [reply tests/create-projects/jenkinsfile_test.go]: Done

File: tests/create-projects/create-cd-jenkins_test.go [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		t.Fatalf(
```
[the remark this answers is not in the document]
Comment (hugowschneider) [reply tests/create-projects/create-cd-jenkins_test.go]: Done

File: jenkins/ocp-config/Tailorfile (hunk 16 above)
[the remark this answers is not in the document]
Comment (hugowschneider) [reply jenkins/ocp-config/Tailorfile]: Reverted

**Comment by hugowschneider** [comment]
@clemensutschig and @michaelsauter 

Good news!!! Github actions solved the problem :)
https://github.com/BIX-Digital/ods-core/commit/b704b134a824dccc89bbcce033b7250b8c00abc2/checks?check_suite_id=392516816

I can run everything using GitHub's actions. Now we need to set up actions in this repository.

**Comment by clemensutschig** [comment]
Love it.. do you close #342 then?

**Comment by hugowschneider** [comment]
There is a little detail that needs investigation. But it should not block this PR. 

`sudo service docker restart` sometimes fails. My guess is that the subnet 172.17.0.1/16 sometimes conflicts with the VM subnet.

**Comment by hugowschneider** [comment]
This PR depends on #357

**Comment by metmajer** [comment]
Love this. Great work! 🥇

**Pull request merged**